uaclient/entitlements/tests/test_livepatch.py

```diff
@@ -0,0 +1,204 @@
+
+        # Make livepatch resource access report not entitled
+        livepatch_unentitled['entitlement']['entitled'] = False
+        self.cfg.write_cache('machine-access-livepatch', livepatch_unentitled)
+        self.assertEqual(status.NONE, self.entitlement.contract_status())
+
+
+class TestLivepatchOperationalStatus(BaseEnabledLivepatchConfig):
+
+    def test_operational_status_inapplicable_on_checked_affordances(self):
+        """The operational_status details failed check_affordances."""
+        livepatch_bionic = copy.deepcopy(dict(LIVEPATCH_RESOURCE_ENTITLED))
+        livepatch_bionic['entitlement']['affordances']['series'] = ['bionic']
+        self.cfg.write_cache('machine-access-livepatch', livepatch_bionic)
+
+        with mock.patch('uaclient.util.get_platform_info') as m_platform_info:
+            m_platform_info.return_value = 'xenial'
+            op_status, details = self.entitlement.operational_status()
+        assert op_status == status.INAPPLICABLE
+        assert 'Livepatch is not available for Ubuntu xenial.' == details
+
+    def test_contract_status_unentitled(self):
+        """The contract_status returns NONE when entitled is False."""
+        self.cfg.write_cache('machine-access-livepatch',
+                             {'entitlement': {'entitled': False}})
+        self.assertEqual(status.NONE, self.entitlement.contract_status())
+
+
+class TestLivepatchProcessDirectives(TestCase):
+
+    @params(('remoteServer', 'remote-server=%s'), ('caCerts', 'ca-certs=%s'))
+    def test_process_directives_call_livepatch_config_command(
+            self, directive_key, livepatch_param_tmpl):
+        """Livepatch config directives are passed to livepatch config."""
+        directive_value = '%s-value' % directive_key
+        cfg = {'entitlement': {'directives': {directive_key: directive_value}}}
+        with mock.patch('uaclient.util.subp') as m_subp:
+            process_directives(cfg)
+        expected_subp = mock.call(
+            ['/snap/bin/canonical-livepatch', 'config',
+             livepatch_param_tmpl % directive_value], capture=True)
+        assert [expected_subp] == m_subp.call_args_list
+
+    def test_process_directives_handles_multiple_directives(self):
+        """Handle multiple Livepatch directives using livepatch config."""
+        cfg = {
+            'entitlement': {'directives': {'remoteServer': 'value1',
+                                           'caCerts': 'value2',
+                                           'ignored': 'ignoredvalue'}}}
+        with mock.patch('uaclient.util.subp') as m_subp:
+            process_directives(cfg)
+        expected_calls = [
+            mock.call(['/snap/bin/canonical-livepatch', 'config',
+                       'remote-server=value1'], capture=True),
+            mock.call(['/snap/bin/canonical-livepatch', 'config',
+                       'ca-certs=value2'], capture=True)]
+        assert expected_calls == m_subp.call_args_list
```

Comment:
Thanks for adding the test!

Given that the order of these calls doesn't matter, we could make this `assert set(expected_calls) == set(m_subp.call_args_list)` which would make this test a little less sensitive to future changes.

(Or, really, if we're doing that then we should just construct `expected_calls` as a set.)

Reply:
I appreciate the suggestion, though I can't cast mock.call to a set  because: 
```
    'ca-certs=value2'], capture=True)])
TypeError: unhashable type: '_Call'
````

Normally I'd agree that the unit test should be flexible regarding ordering (especially when dealing with py2 vs py3  ordering of dict.keys()) but in this case process_directives explicitly processes remoteServer key before caCerts, so the test is correct. 

I'll look around for an alternative for handling the set cast of mock.call. But otherwise I'd be for passing on this suggestion.

Reply:
We've decided it's not worth in on this case but will attempt to use the approach for other list types where ordering doesn't matter.